re-organised code structure and added some vector operation functions

dylandy opened the pull request:
Also I added test build and format checking to travis.yml to make sure the whole thing works.

dylandy committed:
added two iterators for vector and two basic math operation to vector operation

dylandy committed:
added travis yml to project and reformated files

dylandy committed:
fixed require error for iterator file

dylandy committed:
added libgsl dependencies to travis ci

dylandy committed:
try other package for libgsl-dev

dylandy committed:
changed shard link to my github repo

dylandy committed:
added basic calculation to vector and added basic size validator for calculation using and added lengtherror for validation

dylandy committed:
added dependencies to README and shard.yml

dylandy committed:
removed crystal-icr from shard and added crystal-icr repo link to READMe

dylandy committed:
added fedora gsl dependencies

dylandy committed:
added inner product function to vector

dylandy committed:
added proportion methods to vector and added some basic operations to vector

dylandy committed:
fixed format

dylandy committed:
added vector declaration function to README

jlebray committed:
Fix method call (#2)

dylandy committed:
added inspect for vector to show the content

dylandy committed:
added shape method to function

dylandy committed:
replaced original vector operation from stratch to LibGSL function

dylandy committed:
Merge branch 'master' into dev

dylandy committed:
fixed format problem

dylandy committed:
removed test for travis

ruivieira commented:
@dylandy, thanks for your work. Looks good.

Just some things before merging:

- Reorganising the structure means that the specs are broken, since `spec/spec_helper.cr` has the wrong paths (also is `crystal spec` removed from the CI?)
- `shard.yml` (and the README.md) is not pointing to this repo
- GSL is quite big, so I think it's not feasible to put examples and docs in the README.md. A GitHub pages doc site can be set up.
- Is there a reason for `Vector` to be spread over two files (`gsl/base/vector.cr` and `gsl/maths/basic/vector.cr`?

Thanks.

dylandy committed:
removed duplicated push method for vector

dylandy committed:
moved sorting to maths/basic and fixed the path of spec_helper

dylandy committed:
removed usage example from README

dylandy committed:
modified README for shards.yml from dylandy/crystal-gsl to ruivieira/crystal-gsl

dylandy commented:
* I think it is a better way to branch out a new ```test``` branch for spec files so that I removed spec script from travis.yml. Maybe we can put the spec script on the ```test``` branch.

* fixed ```shard.yml``` repo problem in  ```README.md``` 

* Removed usage example from ```README.md```

* I removed the duplicated code in those files. The reason I separated the same module in different location is trying to make task clear. Basic object operation I would put it in ```gsl/base/vector.cr``` and the operations that do things about maths, I would put in ```gsl/maths/basic/vector.cr```

Sorry for so many mistake. Orz

ruivieira commented:
@dylandy Thanks. Looking good.

> I think it is a better way to branch out a new test branch for spec files so that I removed spec script from travis.yml. Maybe we can put the spec script on the test branch.

That's fine, as long we get checks for the PRs. We can always change this in the future.

> I removed the duplicated code in those files. The reason I separated the same module in different location is trying to make task clear. Basic object operation I would put it in gsl/base/vector.cr and the operations that do things about maths, I would put in gsl/maths/basic/vector.cr

That's a good idea.

Regarding the overall code structure of the library, I'll open a comment issue where we call discuss it and then eventually freeze any changes for the first release.
